### PR TITLE
add eddy4R storage workflow Rmd

### PR DESCRIPTION
**Pull request by kexu2014**

### CODE REVIEW

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 1]
```diff
@@ -0,0 +1,2209 @@
...
+    "co2Stor" = c(30) * 60, #output node in dp01
+    "h2oStor" = c(30) * 60,  #output node in dp01
+    "tempStor" = c(30) * 60  #output node in dp01
+    
+    
+  )
+  
+  Para$Flow$dp02 <- list()
+  
+  Para$Flow$dp02$LvlTowrReso <- list(
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: suggest to add detail

**Comment by mjones01** [comment]
@cklunch Is this PR part of the current version of the eddy4R module?

**Comment by cklunch** [comment]
@mjones01 This is a separate vignette, specifically for the storage component. The first vignette is about the turbulent calculations.

**Comment by mjones01** [comment]
@kexu2014 @ddurden @cklunch - Hi.  Is this a vignette that is being worked on or should be updated and then released?  Thanks

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 2]
```diff
@@ -0,0 +1,2209 @@
...
+syntax to explicitly refer to functions. This is done to avoid any 
+possibility of ambiguity, in case there are functions in 
+different packages with the same name.
+
+
+
+```{r pack-install,  message=F, warning=F}
+
+
+install.packages("memisc")
```
[the remark this answers is not in the document]
Comment (kexu2014) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: revised.

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 3]
```diff
@@ -0,0 +1,2209 @@
...
+Here we set file paths and basic input parameters as environment 
+variables for convenience.
+
+
+```{r env-vars}
+
+# define global environment
+eddy4R.base::def.env.glob()
+
+# these parameters need to go into Tammy's environment file
```
[the remark this answers is not in the document]
Comment (kexu2014) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: deleted the comment

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 4]
```diff
@@ -0,0 +1,2209 @@
...
+
+# directory for hdf5 dp0p input data files in the Docker-internal directory structure (char)
+if(FALSE) {
```
[the remark this answers is not in the document]
Comment (kexu2014) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: Revised.

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 5]
```diff
@@ -0,0 +1,2209 @@
...
+# check environment variables for eddy4R workflow parameter "DirFilePara"
+if("METHPARAFLOW" %in% base::names(base::Sys.getenv())) {
+  
+  Para$Flow <- eddy4R.base::def.para.flow(MethParaFlow = "EnvVar")
+  
+  # in case no dp0p h5 file is specified for parameters, download gold files and assign corresponding DirFilePara
+} else {
+  
+  #The dates for the run must be specified, this set is for the gold file
+  DateOut<- "2017-05-23" #Should talk to Stefan about best way to handle this
```
[the remark this answers is not in the document]
Comment (kexu2014) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: deleted the comment

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 6]
```diff
@@ -0,0 +1,2209 @@
...
+allotment.
+
+```{r ingest,  message=F, warning=F}
+
+library(memisc)
+
+numDate <- 0
+tmpOut <- list()
+# assign list
+DATA <- list()
```
[the remark this answers is not in the document]
Comment (kexu2014) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: Which one does not follow eddy4r term? tmpOut? numDate or DATA? or all of them?

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 7]
```diff
@@ -0,0 +1,2209 @@
...
+    "PrdMeas" = 3,
+    "lvlCrdH2oValvVali" = Para$Flow$dp01$LvlTowr$crdH2oValvVali
+  )
+  
+  
+  
+  #Input parameter for qfqm in dp01
+  Para$Flow$dp01$Inp$qfqm <- list()
+  
+  for (nn in names(Para$Flow$dp01$Inp$data)){
```
[the remark this answers is not in the document]
Comment (kexu2014) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: nn is like temporary index in this for loop. changed to idxTmp

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 8]
```diff
@@ -0,0 +1,2209 @@
...
+    Para$Flow$dp01$Inp$qfqm[[nn]]$samp$lvlValvAux <- NULL
+    Para$Flow$dp01$Inp$qfqm[[nn]]$vali$lvlValvAux <- NULL
+  }
+
```
[the remark this answers is not in the document]
Comment (kexu2014) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: This is for qfqm part. I am not sure what they mean. If you feel these lines are unnecessary, we can delete them.

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 9]
```diff
@@ -0,0 +1,2189 @@
...
+if(length(packFlow01) > 0) {
+    update.packages(ask = FALSE, checkBuilt = TRUE)
+    utils::install.packages(packFlow01)
+}
+rm(packFlow00, packFlow01)
+
+# load and attach packages
+library(DataCombine)
+library(eddy4R.base)
+library(eddy4R.turb)
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: may need to add eddy4R.stor

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 10]
```diff
@@ -0,0 +1,2189 @@
...
+DespLoca$var$crdH2o <- c("rtioMoleWetH2o", "rtioMoleDryH2o", "dlta18OH2o",     "dlta2HH2o", "temp",  "pres")
+DespLoca$var$crdH2oValvVali <- "injNum"
+DespLoca$var$envHut <- c("rtioMoleWetH2o", "rh",             "temp" ,          "pres"   )
+
+# map parameters (workflow and scientific) to internal process
+
+
+
+#freq output for all data products
+#TODO(Done): unit should be in [s]
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: This comment should remove

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 11]
```diff
@@ -0,0 +1,2189 @@
...
+             "crdCo2", "crdH2o", "crdH2oValvVali", "crdCo2ValvLvl", "crdH2oValvLvl", "envHut", "valvAux")
+
+#define dp0p sensor names which related to dp01 to dp04 generation
+Sens <-c()
+for (i in 1:length(inpSens)){
+  if((paste0("/", Para$Flow$Loc, "/dp0p/data/", inpSens[i]) %in% listPara$group)) Sens[i] <- inpSens[i]
+}
+
+# principal sampling frequency, used for resampling
+# TODO (Done): incorporate in Para list
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: remove TODO comments

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd (hunk 6 above)
[the remark this answers is not in the document]
Comment (NDurden) [reply tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: DATA

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 12]
```diff
@@ -0,0 +1,2189 @@
...
+
+print(paste0(format(Sys.time(), "%F %T"), ": dataset ", Date, " dp01 calculation begin"))
+
+# assign list for masked input data and parameters
+wrk <- list()
+
+#determine which entries in DATA correspond to desired aggregation interval 
+
+
+#TODO(Done): apply def.idx.agr.r to replace line 544 to line 550
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: remove TODO

**Comment by kexu2014** [comment]
@NDurden Hi, Natchaya,

I revised the vignette based on all of your comments from above and from our previous chat. Just I did not find a way to change the format of #3. Looks like turb vignette has the same problem. But please let me know your thoughts based on the current change.  if further changes needed, I will work on it next weekend. Thanks!

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 13]
```diff
@@ -0,0 +1,1756 @@
...
+# check environment variables for eddy4R workflow parameter "DirFilePara"
+if("METHPARAFLOW" %in% base::names(base::Sys.getenv())) {
+
+  Para$Flow <- eddy4R.base::def.para.flow(MethParaFlow = "EnvVar")
+
+  # in case no dp0p h5 file is specified for parameters, download gold files and assign corresponding DirFilePara
+} else {
+
+  #The dates for the run must be specified, this set is for the gold file
+  DateOut<- "2018-11-28" #Should talk to Stefan about best way to handle this
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: comment need to delete

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 14]
```diff
@@ -0,0 +1,1756 @@
...
+    library(x, character.only = TRUE)
+  }})
+#Remove workflow package list
+rm(packReq)
+
+# load and attach packages
+library(DataCombine)
+library(eddy4R.base)
+library(eddy4R.turb)
+library(ff)           # need to be loaded and attached for arithmetic to work properly on ff objects
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: I believe this workflow did not use ff object thus line 62-67 should be removed

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 15]
```diff
@@ -0,0 +1,1756 @@
...
+#indicating which data will be grab
+if (Para$Flow$Deve == TRUE) {
+  DpName <- c("tempAirLvl", "tempAirTop", "fluxHeatSoil", "radiNet", "presBaro") } else {
+    DpName <- c("tempAirLvl", "tempAirTop", "fluxHeatSoil", "radiNet", "tempSoil", "h2oSoilVol", "presBaro")
+  }
+
+#list all group hdf5 path before adding the exixting data
+listParaPre <- rhdf5::h5ls(Para$Flow$DirFilePara, datasetinfo = FALSE)
+
+#get the Para$Flow$dp01Exis$Lvl for each dp
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: change dp to data product

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 16]
```diff
@@ -0,0 +1,1756 @@
...
+  tmpLvl <- grep(pattern = "30m", x = tmpLvl, ignore.case = TRUE, value =  TRUE)
+  tmpLoc <- strsplit(as.character(tmpLvl),split='_', fixed=TRUE)
+  Para$Flow$dp01Exis$Lvl[[DpName[i]]] <- unlist(lapply(1:length(tmpLoc), function(x) {as.character(paste0(tmpLoc[[x]][1], "_", tmpLoc[[x]][2]))}))
+}
+
+rlog$debug("Begin to re-ingest the existing dp01 data")
+base::invisible(eddy4R.base::wrap.hdf5.wrte.dp01.api(date = Para$Flow$DateOut, SiteLoca = Para$Flow$Loc, FileOut = Para$Flow$DirFilePara, LvlTowr = Para$Flow$dp01Exis$Lvl, DpName = DpName))
+rlog$info("End of re-ingest the existing dp01 data")
+# workflow parameters
+
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: please remove non-necessary lines

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 17]
```diff
@@ -0,0 +1,1756 @@
...
+} else {
+
+  dir.create(Para$Flow$DirWrk, recursive = TRUE, showWarnings = FALSE)
+
+}
+
+# set working directory
+setwd(Para$Flow$DirWrk)
+
+#  DirInp, DirMnt input directory
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: suggest to change to # input directory

File: tutorials-in-development/eddy4r/stor.neon.vignette.Rmd [hunk 18]
```diff
@@ -0,0 +1,1756 @@
...
+  Para$Flow$DirOut <- paste0(Para$Flow$DirWrk, "/out")
+
+  # option: create user-specified input directory (if not existing), e.g. on host filesystem
+} else {
+
+  dir.create(Para$Flow$DirOut, recursive = TRUE, showWarnings = FALSE)
+
+}
+
+
```
Comment (NDurden) [inline tutorials-in-development/eddy4r/stor.neon.vignette.Rmd]: suggest to add #get site name

**Comment by kexu2014** [comment]
@mjones01 
Dear Megan, 

After a couple rounds of review and revise between Natchaya @NDurden and me, this pull request is finally ready to be merged and released. Please let us know if there is any question when merging in this pull request. Thank you!

**Comment by mjones01** [comment]
@kexu2014 Thank you for all the work on this vignettes.

**Pull request merged**